File: ck-android/CkClient/iosApp/iosApp/vpn_service/OutlineDeviceFacadeImpl.swift
import app
import MyLibrary

// IS NOT USED NOW. TEST IMPLEMENTATION
class OutlineDeviceFacadeImpl: OutlineDeviceFacade {
    
    private var deviceFacade = DeviceFacade()
    
    func doInit(apiKey: String) {
        deviceFacade.initialize(config: apiKey)
    }
    
    func read() -> KotlinByteArray? {
        let data: Data = deviceFacade.read()
        return KotlinByteArray.from(data: data)
    }
    
    func write(data: KotlinByteArray) {
        let sz = Int32(data.size)
        var uintArray: [UInt8] = []
        
        for i in 0..<sz {
            var byte = data.get(index: i)
            uintArray.append(UInt8(byte))
        }
                
        var convertedData = convertKotlinByteArrayToData(kotlinByteArray: uintArray)
        deviceFacade.write(data: convertedData)
    }
    
    func convertKotlinByteArrayToData(kotlinByteArray: [UInt8]) -> Data {
        return Data(kotlinByteArray)
    }
}

extension KotlinByteArray {
    static func from(data: Data) -> KotlinByteArray {
        let int8array = [UInt8](data)
            .map(Int8.init(bitPattern:))
        
        let result = KotlinByteArray(size: Int32(data.count))
        for i in 0..<data.count {
            result.set(index: Int32(i), value: int8array[i])
        }
        
        return result
    }
}

File: ck-android/CkClient/app/src/androidMain/kotlin/com/dobby/feature/vpn_service/domain/OutlineDeviceFacadeImpl.kt
package com.dobby.feature.vpn_service.domain

import cloak_outline.Cloak_outline
import cloak_outline.OutlineDevice
import com.dobby.feature.vpn_service.OutlineDeviceFacade

class OutlineDeviceFacadeImpl: OutlineDeviceFacade {

    private var device: OutlineDevice? = null

    override fun init(apiKey: String) {
        device = Cloak_outline.newOutlineDevice(apiKey)
    }

    override fun write(data: ByteArray) {
        device?.write(data)
    }

    override fun read(): ByteArray? {
        return device?.read()
    }
}
